Revamp build history search bar

janfaracik opened the pull request:
### Proposed changelog entries

* Modernise the "Build History" search bar

### Video

https://user-images.githubusercontent.com/43062514/130856620-48c47dc0-4385-4313-8194-050ec0ad07ba.mov


### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

janfaracik committed:
Redesign search builds field

janfaracik committed:
initial

janfaracik committed:
Update icons

janfaracik committed:
Update index.jelly

janfaracik committed:
Update style.less

janfaracik committed:
Update style.less

janfaracik committed:
Remove unused things

janfaracik committed:
Merge master

janfaracik committed:
Fix icons

janfaracik committed:
Remove unused icons

janfaracik committed:
Update IconSet.java

janfaracik committed:
Update index.jelly

janfaracik committed:
Rework search

Wadeck reviewed (commented):
Not manually tested but looks really nice :)

fqueiruga commented on war/src/main/webapp/scripts/hudson-behavior.js:
```diff
@@ -2292,7 +2291,7 @@ function updateBuildHistory(ajaxUrl,nBuild) {
 
         function loadPage(params, focusOnSearch) {
-            var searchString = pageSearchInput.value;
+            const searchString = pageSearchInput.value;
```
Cannot use `const` on the `hudson-behavior.js` file as it's not processed through webpack. These should be `var`s

janfaracik replied:
Done.

fqueiruga commented on war/src/main/webapp/scripts/hudson-behavior.js:
```diff
@@ -2342,18 +2335,9 @@ function updateBuildHistory(ajaxUrl,nBuild) {
-        });
-        pageSearchClear.observe('click', function() {
-            pageSearchInput.value = '';
+        pageSearchInput.oninput = function() {
```
I'd suggest to revert this. This would cause a lot of requests to the backend, and there's no sane way of adding a debounce to this file. Other option would be to extract all this functionality to a webpack entrypoint and use `lodash.debounce`

janfaracik replied:
I've extracted the build history JS from `hudson-behavior.js` and added debounce - the file itself is a little rough around the edges at the moment but its working well from the look of it.

fqueiruga replied:
Ok, sounds good

fqueiruga commented:
Looks good. I do have some comments about some issues that will arise.

I'd also suggest using, if possible, the `.setting-input` CSS class to the input to provide the basic styles, and only add the needed styles on top. That way it will be updated if forms input ever get revamped.

oleg-nenashev commented:
@janfaracik Looks like this one is almost ready to go. WDYT about the comments by @fqueiruga ?

janfaracik committed:
Update hudson-behavior.js

janfaracik commented:
> @janfaracik Looks like this one is almost ready to go. WDYT about the comments by @fqueiruga ?

They sound sensible to me 👍 I'll be making some changes, I'll post another comment when it should be good for another look over.

janfaracik committed:
Rough but working commit, rip out build history from hudson-behavior into separate file

janfaracik committed:
Update form.less

janfaracik committed:
Update side-panel-widgets.less

janfaracik committed:
Update filter-build-history.js

janfaracik commented:
I've updated the PR based on the reviews - thanks everybody by the way :) 

I've added debounce so server calls are limited, there's a short delay after typing before calling the API so it shouldn't swamp the API. I've also added a small animation to indicate that it's loading and I've also a small notice when there are no builds.

https://user-images.githubusercontent.com/43062514/133934061-47a4e67d-1e67-4306-bcc7-40a6efb9e96b.mov

Reviews are welcome, I've extracted the JS into its own file now which makes things cleaner and easier to maintain - the file could definitely be cleaner though so would be great for people to try out the PR/suggest improvements.

Cheers

timja commented:
Wow I just tried out the old behaviour and that is not a nice UX, never used this before tbh

timja commented:
Icon appears to be mis-aligned for me?

Running chrome on mac big sur

![image](https://user-images.githubusercontent.com/21194782/133935233-650b2018-4346-46e7-b7c6-304fcb7669ba.png)

timja commented:
Spacing appears to be off for me too, it doesn't appear like in the video above.

The `x` button doesn't change the cursor to show it's clickable.

doesn't look the best in dark theme but looks like there's variables there that can be used to tweak that

janfaracik committed:
Fix styling in Chrome (and also Safari...)

janfaracik commented:
> Spacing appears to be off for me too, it doesn't appear like in the video above.
> 
> The `x` button doesn't change the cursor to show it's clickable.
> 
> doesn't look the best in dark theme but looks like there's variables there that can be used to tweak that

Darn Safari handles padding/line height _**differently**_ for search inputs :doh: Should be all fixed now :) 

The clear button also now changes the cursor on hover.

timja reviewed (approved):
the focus appears to be quite strong (and different from the rest of the application). Not that I like the focus state in the rest of the application.

I think in this case it looks fine, although if it was rolled out more widely it would probably be too 'much' when there's a number of inputs together, but I could be wrong.

My previous feedback was addressed and this is a **lot** better than the previous search bar

timja commented:
conflicts

janfaracik committed:
Merge branch 'master' into search-bar

timja commented:
@fqueiruga any chance you could take another look?

fqueiruga reviewed (approved):
I think this is a great PR. I'm still a bit concious about the use of `px` but that's not a blocker.

Cheers @janfaracik , this is really nice.

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

dheerajodha commented:
Suggested Edit:

- Modernise the <code>Build History</code> search bar.

timja commented:
Possibly caused https://issues.jenkins.io/browse/JENKINS-66753

janfaracik commented:
> Possibly caused https://issues.jenkins.io/browse/JENKINS-66753

Investing this now...
